### PR TITLE
chore: caching

### PR DESCRIPTION
**Pull request by tokers**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [ ] Bugfix
- [x] New feature provided
- [ ] Improve performance
- [ ] Backport patches

- Related issues

___
### Bugfix
- Description

- How to fix?

___
### New feature or improvement
- Describe the details and related test reports.

This PR optimized the caching use in ingress controller, the caching system is hidden inside the `apisix.APISIX` implementation, callers even don't know the existence of caching, so many statements like "first creating, then inserting to cache", "first deleting from APISIX, then deleting from cache" are removed.

The cacheing interface was designed as:

```go
// Cache defines the necessary behaviors that the cache object should have.
// Note this interface is for APISIX, not for generic purpose, it supports
// standard APISIX resources, i.e. Route, Upstream, Service and SSL.
// Cache implementations should copy the target objects before/after read/write
// operations for the sake of avoiding data corrupted by other writers.
type Cache interface {
	// InsertRoute adds or updates route to cache.
	InsertRoute(*v1.Route) error
	// InsertService adds or updates service to cache.
	InsertService(*v1.Service) error
	// InsertSSL adds or updates ssl to cache.
	InsertSSL(*v1.Ssl) error
	// InsertUpstream adds or updates upstream to cache.
	InsertUpstream(*v1.Upstream) error

	// GetRoute finds the route from cache according to the primary index.
	GetRoute(string) (*v1.Route, error)
	// GetService finds the service from cache according to the primary index.
	GetService(string) (*v1.Service, error)
	// GetSSL finds the ssl from cache according to the primary index.
	GetSSL(string) (*v1.Ssl, error)
	// GetUpstream finds the upstream from cache according to the primary index.
	GetUpstream(string) (*v1.Upstream, error)

	// ListRoutes lists all routes in cache.
	ListRoutes() ([]*v1.Route, error)
	// ListServices lists all services in cache.
	ListServices() ([]*v1.Service, error)
	// ListSSL lists all ssl objects in cache.
	ListSSL() ([]*v1.Ssl, error)
	// ListUpstreams lists all upstreams in cache.
	ListUpstreams() ([]*v1.Upstream, error)

	// DeleteRoute deletes the specified route in cache.
	DeleteRoute(*v1.Route) error
	// DeleteService deletes the specified service in cache.
	DeleteService(*v1.Service) error
	// DeleteSSL deletes the specified ssl in cache.
	DeleteSSL(*v1.Ssl) error
	// DeleteUpstream deletes the specified upstream in cache.
	DeleteUpstream(*v1.Upstream) error
}
```

And we have an implementation with the help of go-memdb.

Also, controller will try to sync (with an exponential back off strategy) to the APISIX cluster when the cluster was just added. All operations will fail if the cluster is not ready (the initial sync is failure), we'll add a recover mechanism once the APISIX cluster recovered, the controller shall detect it.
___
### Backport patches
- Why need to backport?

- Source branch

- Related commits and pull requests

- Target branch

### CODE REVIEW

**Comment by tokers** [comment]
@gxthrj Please take a look when you have time.

File: pkg/apisix/cluster.go [hunk 1]
```diff
@@ -85,20 +89,129 @@ func newCluster(o *ClusterOptions) (Cluster, error) {
...
+func (c *cluster) Ready(ctx context.Context) error {
+	if c.cacheWarmingUpErr != nil {
+		return c.cacheWarmingUpErr
+	}
+	select {
+	case <-ctx.Done():
+		log.Errorf("failed to wait cluster to ready: %s", ctx.Err())
+		return ctx.Err()
+	case <-c.cacheReady:
+		return nil
```
Comment (gxthrj) [inline pkg/apisix/cluster.go:211]: Had better to record log when ready .

File: pkg/apisix/route.go [hunk 2]
```diff
@@ -48,9 +48,17 @@ func newRouteClient(c *cluster) Route {
 	}
 }
 
+// Get only looks up the cache, it's not necessary to access APISIX, since all resources
+// are created by Create, which reflects the change to cache in turn, so if resource
+// is not in cache, it's not in APISIX either.
+func (r *routeClient) Get(_ context.Context, fullname string) (*v1.Route, error) {
+	return r.cluster.cache.GetRoute(fullname)
```
Comment (gxthrj) [inline pkg/apisix/route.go:55]: If cache return nil, need to read from admin api.
Or we need a way to resync from APISIX to cache.

File: pkg/seven/state/builder.go [hunk 3]
```diff
@@ -227,64 +221,18 @@ func SolverSingleUpstream(u *v1.Upstream, swg ServiceWorkerGroup, wg *sync.WaitG
...
-						errNotify = err
-						return
-					}
-
-					// 2.sync apisix
-					var cluster string
-					if u.Group != nil {
-						cluster = *u.Group
-					}
+				if needToUpdate || (u.FromKind != nil && *u.FromKind == WatchFromKind) {
```
Comment (gxthrj) [inline pkg/seven/state/builder.go:224]: Need to differentiate `service registration` and `CRD synchronization`, 
When `FromKind == WatchFromKind`, the upstream info only contains nodes, and will lose other information in upstream, such as loadBalance infomation.

File: pkg/seven/state/solver.go [hunk 4]
```diff
@@ -57,92 +57,58 @@ func (s *ApisixCombination) Solver() (string, error) {
...
+		var cluster string
+		if ups.Group != nil {
+			cluster = *ups.Group
 		}
-		return err
-	} else {
-		up.ID = upstream.ID
-	}
-	serviceRequest := db.ServiceRequest{UpstreamId: *up.ID}
-	if svc, err := serviceRequest.ExistByUpstreamId(); err != nil {
```
Comment (gxthrj) [inline pkg/seven/state/solver.go:124]: Where is this logic in this refactor, need to check if upstream is refer by any service.

File: pkg/apisix/cluster.go (hunk 1 above)
Comment (tokers) [reply pkg/apisix/cluster.go:211]: OK.

File: pkg/seven/state/solver.go (hunk 4 above)
Comment (tokers) [reply pkg/seven/state/solver.go:124]: In Delete implementations.

**Comment by gxthrj** [comment]
Unit-test CI  failed @tokers Please have a look.

**Comment by tokers** [comment]
> Unit-test CI failed @tokers Please have a look.

Fixed.

File: pkg/seven/state/builder.go (hunk 3 above)
Comment (tokers) [reply pkg/seven/state/builder.go:235]: Fixed.

File: pkg/apisix/route.go (hunk 2 above)
Comment (tokers) [reply pkg/apisix/route.go]: Added.

File: pkg/apisix/apisix.go [hunk 5]
```diff
@@ -73,39 +81,28 @@ type Upstream interface {
...
 
-// NewForOptions creates an APISIX client to perform resources change pushing.
-// Users should carry a ClusterOptions to configure the default APISIX cluster.
-func NewForOptions(co *ClusterOptions) (APISIX, error) {
-	defaultCluster, err := newCluster(co)
-	if err != nil {
-		return nil, err
-	}
+// New creates an APISIX client to perform resources change pushing.
+func New() (APISIX, error) {
```
Comment (gxthrj) [inline pkg/apisix/apisix.go:97]: Suggest to use `Client()` instead of `New()`

Comment (tokers) [reply pkg/apisix/apisix.go]: Changed to NewClient.

**Review by gxthrj — APPROVED** [review]

**Pull request merged**